### PR TITLE
[FLIRT] Refactoring and create rz_sign library

### PR DESCRIPTION
**Pull request by wargio**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

1. I have moved the `flirt.c` file under `librz/signature/` and created a new lib called `rz_sign` which is used by RzCore & RzMain
2. Small refactoring to remove dead code & global vars from `flirt.c`
3. Moved the `rz_sign_flirt_dump` under `librz/core/csign.c` to avoid the usage of `analysis->cb_printf`
4. Moved sign.c to `librz/signature/zignature.c`
5. Removed usage of core and using RzCoreBind (which will need to be removed in the future before rizin v1.0.0)

The objective is to set the base for the new work on FLIRT

### CODE REVIEW

File: librz/core/cflirt.c [hunk 1]
```diff
@@ -0,0 +1,114 @@
+// SPDX-FileCopyrightText: 2021 RizinOrg <info@rizin.re>
+// SPDX-FileCopyrightText: 2021 deroad <wargio@libero.it>
+// SPDX-FileCopyrightText: 2014-2016 jfrankowski <jody.frankowski@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+#include <rz_core.h>
```
Comment (XVilka) [inline librz/core/cflirt.c]: Put an empty line between comment and include please.

**Review by XVilka — CHANGES_REQUESTED** [review]
I don't think we should move FLIRT in its own library. I think it should be a more generic `librz/signature/` or something. And later move some of the code from `librz/analysis/sign.c` or whatever new signature code will be.

The same approach that we did with RZIL - core code is in `librz/il` and analysis-dependent code in `librz/analysis/rzil`, and core-dependent code in `librz/core/cil.c`.

File: .github/labeler.yml [hunk 2]
```diff
@@ -128,6 +128,11 @@ RZIL:
   - librz/il/**/*
   - librz/core/cil.c
 
+FLIRT:
+  - librz/flirt/*
```
Comment (XVilka) [inline .github/labeler.yml]: Please fix this as well then. Also, add or update the `signature` label as well.

File: librz/core/cmd/cmd_zign.c [hunk 3]
```diff
@@ -544,7 +545,7 @@ static int cmdFlirt(void *data, const char *input) {
 			eprintf("Usage: zfd filename\n");
 			return false;
 		}
-		rz_sign_flirt_dump(core->analysis, input + 2);
+		rz_core_flirt_dump(input + 2);
```
Comment (XVilka) [inline librz/core/cmd/cmd_zign.c:548]: In the future, we should split opening and parsing, and printing information from the parsed FLIRT file.

File: librz/core/cmd/cmd_zign.c [hunk 4]
```diff
@@ -557,7 +558,7 @@ static int cmdFlirt(void *data, const char *input) {
 		RzListIter *iter;
 		RzList *files = rz_file_globsearch(input + 2, depth);
 		rz_list_foreach (files, iter, file) {
-			rz_sign_flirt_scan(core->analysis, file);
+			rz_flirt_apply_signatures(core->analysis, file);
```
Comment (XVilka) [inline librz/core/cmd/cmd_zign.c]: Rename to `rz_sign_flirt_apply()` or something similar.

File: librz/signature/flirt.c [hunk 5]
```diff
@@ -347,49 +315,45 @@ ut16 crc16(const unsigned char *data_p, size_t length) {
 	return (ut16)(crc);
 }
 
-// this is ugly, but we can't afford to change the return size of read_byte
-static bool buf_eof;
```
Comment (XVilka) [inline librz/signature/flirt.c:351]: Good catch!

File: librz/signature/flirt.c [hunk 6]
```diff
@@ -409,120 +373,23 @@ static void module_free(RzFlirtModule *module) {
...
-		rz_list_free(module->referenced_functions);
-	}
+	rz_list_free(module->public_functions);
+	rz_list_free(module->tail_bytes);
+	rz_list_free(module->referenced_functions);
 	free(module);
 }
 
-static void node_free(RzFlirtNode *node) {
+RZ_API void rz_flirt_node_free(RZ_NULLABLE RzFlirtNode *node) {
```
Comment (XVilka) [inline librz/signature/flirt.c]: Add a Doxygen for the API function.

File: librz/signature/flirt.c [hunk 7]
```diff
@@ -605,7 +472,7 @@ static int module_match_buffer(RzAnalysis *analysis, const RzFlirtModule *module
 			next_module_function->name = rz_str_newf("flirt.%s", name);
 			analysis->flb.set(analysis->flb.f, next_module_function->name,
 				next_module_function->addr, next_module_function_size);
-			analysis->cb_printf("Found %s\n", next_module_function->name);
+			RZ_LOG_INFO("flirt: Found %s\n", next_module_function->name);
```
Comment (XVilka) [inline librz/signature/flirt.c]: Maybe use the original name - `FLIRT` in the error messages and warnings?

File: librz/signature/flirt.c [hunk 8]
```diff
@@ -653,14 +520,15 @@ static int node_match_buffer(RzAnalysis *analysis, const RzFlirtNode *node, ut8
...
+/* Tries to find matching functions between the signature infos in root_node
+ * and the analyzed functions in analysis
+ * Returns false on error. */
+static bool node_match_functions(RzAnalysis *analysis, const RzFlirtNode *root_node) {
+	bool ret = true;
 
 	if (rz_list_length(analysis->fcns) == 0) {
-		analysis->cb_printf("There are no analyzed functions. Have you run 'aa'?\n");
-		return true;
+		RZ_LOG_ERROR("flirt: There are no analyzed functions. Have you run 'aa'?\n");
```
Comment (XVilka) [inline librz/signature/flirt.c]: Same.

File: librz/signature/flirt.c [hunk 9]
```diff
@@ -674,12 +542,14 @@ static int node_match_functions(RzAnalysis *analysis, const RzFlirtNode *root_no
...
 		if (!func_buf) {
-			continue;
+			ret = false;
+			break;
 		}
 		if (!analysis->iob.read_at(analysis->iob.io, func->addr, func_buf, (int)func_size)) {
-			eprintf("Couldn't read function %s at 0x%" PFMT64x "\n", func->name, func->addr);
-			free(func_buf);
-			continue;
+			RZ_LOG_ERROR("flirt: Couldn't read function %s at 0x%" PFMT64x "\n", func->name, func->addr);
```
Comment (XVilka) [inline librz/signature/flirt.c]: Same.

File: librz/signature/flirt.c [hunk 10]
```diff
@@ -1316,41 +1194,40 @@ static RzFlirtNode *flirt_parse(const RzAnalysis *analysis, RzBuffer *flirt_buf)
...
 
-	if (!(version = rz_sign_is_flirt(flirt_buf))) {
+	if (!(ps.version = rz_flirt_get_version(flirt_buf))) {
 		goto exit;
 	}
 
-	if (version < 5 || version > 10) {
-		eprintf("Unsupported flirt signature version\n");
+	if (ps.version < 5 || ps.version > 10) {
+		RZ_LOG_ERROR("flirt: Unsupported flirt signature version\n");
```
Comment (XVilka) [inline librz/signature/flirt.c]: `FLIRT`

File: librz/signature/flirt.c [hunk 11]
```diff
@@ -1372,54 +1249,48 @@ static RzFlirtNode *flirt_parse(const RzAnalysis *analysis, RzBuffer *flirt_buf)
...
 	if (rz_buf_read(flirt_buf, buf, size) != size) {
 		goto exit;
 	}
 
 	if (header->features & IDASIG__FEATURE__COMPRESSED) {
-		if (version >= 5 && version < 7) {
+		if (ps.version >= 5 && ps.version < 7) {
 			if (!(decompressed_buf = rz_inflate_ignore_header(buf, size, NULL, &decompressed_size))) {
-				eprintf("Decompressing failed.\n");
+				RZ_LOG_ERROR("flirt: Failed to decompress buffer.\n");
```
Comment (XVilka) [inline librz/signature/flirt.c]: Same.

File: librz/signature/flirt.c [hunk 12]
```diff
@@ -1372,54 +1249,48 @@ static RzFlirtNode *flirt_parse(const RzAnalysis *analysis, RzBuffer *flirt_buf)
...
 				goto exit;
 			}
-		} else if (version >= 7) {
+		} else if (ps.version >= 7) {
 			if (!(decompressed_buf = rz_inflate(buf, size, NULL, &decompressed_size))) {
-				eprintf("Decompressing failed.\n");
+				RZ_LOG_ERROR("flirt: Failed to decompress buffer.\n");
```
Comment (XVilka) [inline librz/signature/flirt.c]: Same.

File: librz/signature/flirt.c [hunk 13]
```diff
@@ -1372,54 +1249,48 @@ static RzFlirtNode *flirt_parse(const RzAnalysis *analysis, RzBuffer *flirt_buf)
...
 				goto exit;
 			}
 		} else {
-			eprintf("Sorry we do not support the signatures"
-				" version %c compression.\n",
-				version);
+			RZ_LOG_ERROR("flirt: Sorry we do not support compressed signatures with version %d.\n", ps.version);
```
Comment (XVilka) [inline librz/signature/flirt.c]: Same.

File: librz/signature/flirt.c [hunk 14]
```diff
@@ -1456,48 +1338,32 @@ RZ_API int rz_sign_is_flirt(RzBuffer *buf) {
...
+ * \param flirt_file  The FLIRT file to parse
+ */
+RZ_API void rz_flirt_apply_signatures(RzAnalysis *analysis, const char *flirt_file) {
+	rz_return_if_fail(analysis && RZ_STR_ISNOTEMPTY(flirt_file));
+	RzBuffer *flirt_buf = NULL;
+	RzFlirtNode *node = NULL;
 
 	if (!(flirt_buf = rz_buf_new_slurp(flirt_file))) {
-		eprintf("Can't open %s\n", flirt_file);
+		RZ_LOG_ERROR("flirt: Can't open %s\n", flirt_file);
```
Comment (XVilka) [inline librz/signature/flirt.c]: Also `FLIRT`.

File: librz/signature/flirt.c [hunk 15]
```diff
@@ -1456,48 +1338,32 @@ RZ_API int rz_sign_is_flirt(RzBuffer *buf) {
...
 		return;
 	}
 
-	node = flirt_parse(analysis, flirt_buf);
+	node = rz_flirt_parse_buffer(flirt_buf);
 	rz_buf_free(flirt_buf);
 	if (node) {
 		if (!node_match_functions(analysis, node)) {
-			eprintf("Error while scanning the file %s\n", flirt_file);
+			RZ_LOG_ERROR("flirt: Error while scanning the file %s\n", flirt_file);
```
Comment (XVilka) [inline librz/signature/flirt.c]: Same.

File: librz/signature/flirt.c [hunk 16]
```diff
@@ -1456,48 +1338,32 @@ RZ_API int rz_sign_is_flirt(RzBuffer *buf) {
...
 		}
-		node_free(node);
+		rz_flirt_node_free(node);
 		return;
 	} else {
-		eprintf("We encountered an error while parsing the file %s. Sorry.\n", flirt_file);
+		RZ_LOG_ERROR("flirt: We encountered an error while parsing the file %s. Sorry.\n", flirt_file);
```
Comment (XVilka) [inline librz/signature/flirt.c]: Same.

File: librz/signature/zignature.c [hunk 17]
```diff
@@ -2738,3 +2723,34 @@ RZ_API RzSignOptions *rz_sign_options_new(const char *bytes_thresh, const char *
...
+
+static void zign_rename_for(RzEvent *ev, int type, void *user, void *data) {
+	RzSpaces *s = (RzSpaces *)ev->user;
+	RzAnalysis *analysis = container_of(s, RzAnalysis, zign_spaces);
+	RzSpaceEvent *se = (RzSpaceEvent *)data;
+	rz_sign_space_rename_for(analysis, se->data.rename.space,
+		se->data.rename.oldname, se->data.rename.newname);
+}
+
+RZ_API void rz_sign_analysis_set_hooks(RZ_NONNULL RzAnalysis *analysis) {
```
Comment (XVilka) [inline librz/signature/zignature.c:2755]: Add Doxygen, please.

**Review by XVilka — CHANGES_REQUESTED** [review]
Overall very good, just a few minor things to fix first.

File: librz/core/cmd/cmd_zign.c (hunk 3 above)
Comment (wargio) [reply librz/core/cmd/cmd_zign.c:548]: agreed, but this is ok for now.

File: librz/core/cmd/cmd_zign.c (hunk 4 above)
Comment (wargio) [reply librz/core/cmd/cmd_zign.c:561]: why? i think the new name is more clear.

File: librz/core/csign.c [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+	}
+	rz_cons_printf(":\n");
+}
+
+static void flirt_print_indentation(int indent) {
+	rz_cons_printf("%s", rz_str_pad(' ', indent));
+}
+
+static void flirt_print_node(const RzFlirtNode *node, int indent) {
+	/*Prints a signature node. The output is similar to dumpsig*/
```
Comment (XVilka) [inline librz/core/csign.c]: Missing space before and after.

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
@wargio after merge:
```
../librz/include/rz_bind.h:27:23: warning: redefinition of typedef 'RzCoreConfigGet' is a C11 feature [-Wtypedef-redefinition]
typedef const char *(*RzCoreConfigGet)(void *core, const char *key);
                      ^
../librz/include/rz_bind.h:25:23: note: previous definition is here
typedef const char *(*RzCoreConfigGet)(void *core, const char *key);
                      ^
1 warning generated.
```

**Comment by wargio** [comment]
@XVilka https://github.com/rizinorg/rizin/pull/1967

**Pull request merged**